## Exploring European Option Properties and the Black-Scholes Model

### Introduction

In this notebook, we will explore key properties of european options, focusing on both theoretical insights and practical applications. Specifically, we aim to demonstrate important option characteristics and confirm that the Black-Scholes (BS) Model aligns with these properties. Understanding these features can help clarify how options work, why they are valued the way they are, and how they behave under various market conditions.

### Objectives

This notebook is structured to help you:

1. **Understand Core Option Properties**: We will examine the basic properties of european options, including intrinsic value, and how the price of an option behaves like a mathematical function and what are the no arbitrage conditions for this price .  
2. **Verify Black-Scholes Model Conformance**: The Black-Scholes Model is one of the most widely used models for pricing European options. We'll apply the BS Model to our data and validate that it accurately respects the fundamental properties of options, such as:
   - **Put-Call Parity**: This principle establishes a relationship between the prices of European call and put options, ensuring no-arbitrage conditions.
   - **Monotonicity**
   - **No Arbitrage interval range for various Black & Scholes Parameters**

### Demonstration Outline

Throughout this notebook, we will:

- **Apply the Black-Scholes formula** to calculate theoretical european option values and assess their alignment with actual option behaviors.

By the end of this notebook, we will have a deeper understanding of option properties and an empirical view of the Black-Scholes Model's effectiveness in modeling these properties. This hands-on approach will reinforce key concepts in options trading and valuation, as well as demonstrate the practical relevance of financial modeling.

---

Let's get started!


### Call Option

Consider a European call option with:
- Price: $C(S_t, t; K, T)$
- Spot price at time $t:S_t$
- Expiry date: $T$
- Strike price: $K$

### First Property 

The call option price is always greater than or equal to its intrinsic value. That is :

$$
C(S_t, t; K, T) \geq \max(S_t - K, 0)
$$

where:
- $C(S_t, t; K, T)$ is the price of the European call option at time $t$,
- $S_t$ is the current spot price of the underlying asset,
- $K$ is the strike price of the option, and
- $T$ is the expiration date of the option.

### Second Property 

The call option price is always less or equal to the underlying asset price $S_t$ . That is:

$$
C(S_t, t; K, T) \leq S_t
$$


### Consequences :  $$\max(S_t - K, 0) \leq C(S_t, t; K, T) \leq S_t$$

In [1]:
%run Europea_Options.ipynb  # We run our notebook on Black & Scholes Model Class for European Option we previously created

In [2]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

59.10115212246

In [3]:
call_intrinsic_value = max(100-50,0)
call_intrinsic_value

50

$$50 \leq C(100, 0; 50, 1) \leq 100$$  

So, our option will always cost at least 50 dollars and will always be less than 100 dollars . Hence, Black & Scholes model is coherent since it reproduces this property of the call option price and the interval where our call price must fall is \[50,100\].

### Put Option

Consider now a European Put option with:
- Price: $P(S_t, t; K, T)$
- Spot price at time $t:S_t$
- Expiry date: $T$
- Strike price: $K$

### First Property 

The Put option price is always greater than or equal to its intrinsic value. That is :

$$
P(S_t, t; K, T) \geq \max(K - S_t, 0)
$$

where:
- $P(S_t, t; K, T)$ is the price of the European put option at time $t$,
- $S_t$ is the current spot price of the underlying asset,
- $K$ is the strike price of the option, and
- $T$ is the expiration date of the option.

### Second Property 

The put option price is always less or equal to the strike $K$ . That is:

$$
P(S_t, t; K, T) \leq K
$$


### Consequences :  $$\max(K - S_t, 0) \leq P(S_t, t; K, T) \leq K$$

In [4]:
european_option.Put_Black_Scholes()

0.03768977635909404

In [5]:
put_intrinsic_value = max(50-100,0)
put_intrinsic_value

0

$$0 \leq P(100,0;50,1) \leq 50$$

Hence, a put will cost at least 0 dollars ( since the option is currently out of the money ) and will always be less or equal than 50 dollars . Hence, Black & Scholes model is coherent since it reproduces this property of the put option price and the interval where our put price must fall is \[0,50\].

## Call Option 

### Third Property 

Consider two European call options $C(S_t, t; K, T_1)$ and $C(S_t, t; K, T_2)$, with expiry times $T_1$ and $T_2$ such that $T_1 < T_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
C(S_t, t; K, T_1) \leq C(S_t, t; K, T_2)
$$
for $T_1 < T_2$.

In [6]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes() # Maturity Here is 1 year 

59.10115212246

In [7]:
european_option = EuropeanOptions(S_t=100,K=50,T=2,sigma=0.35,r=0.20,D=0) # Maturity of 2 years 
european_option.Call_Black_Scholes()

66.61735526893983

The more the maturity, the more the uncertainty and the more the option price will be !!! So Black & Scholes model is coherent again to take into account this property .

## Put Option 

### Third Property 

Consider two European put options $P(S_t, t; K, T_1)$ and $P(S_t, t; K, T_2)$, with expiry times $T_1$ and $T_2$ such that $T_1 < T_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
P(S_t, t; K, T_1) \leq P(S_t, t; K, T_2)
$$
for $T_1 < T_2$.

In [8]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

0.03768977635909404

In [9]:
european_option = EuropeanOptions(S_t=100,K=50,T=2,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

0.13335757072179677

The more the maturity , the more the uncertainty , the more the put option price !!! The Black & Scholes model is coherent again .

## Call Option 

### Fourth Property

Consider two European call options $C(S_t, t; K_1, T)$ and $C(S_t, t; K_2, T)$ having strike prices $K_1$ and $K_2$, respectively, with $K_1 < K_2$.
$S_t$ is the spot price at time $t$ and $T > t$ is the expiry time.

Then
 
$$
 0 \leq C(S_t, t; K_1, T) - C(S_t, t; K_2, T) \leq K_2 - K_1
$$
 
$K_1 < K_2$.

In [10]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

59.10115212246

In [11]:
european_option = EuropeanOptions(S_t=100,K=75,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

39.604649616580126

In [12]:
59.10115212246 - 39.604649616580126

19.496502505879874

$$0 \leq 19.496502505879874 \leq 25$$
$$0 \leq C(100,0;50,1) - C(100,0;75,1) \leq 25$$

Hence, if we have two call options of diffrent strikes $K_1$ and $K_2$, we can determine the interval range where the difference between the call option strike $K_1$ and the call option strike $K_2$ fall . In no arbitrage opportunity, then, this interval will always be valid. It can be easily used to verify whether we have arbitrage opportunity or not with two different call option strikes .

## Put Option 

### Fourth Property

Consider two European put options $P(S_t, t; K_1, T)$ and $P(S_t, t; K_2, T)$ having strike prices $K_1$ and $K_2$, respectively, with $K_1 < K_2$.
$S_t$ is the spot price at time $t$ and $T > t$ is the expiry time.

Then
 
$$
 0 \leq P(S_t, t; K_2, T) - P(S_t, t; K_1, T) \leq K_2 - K_1
$$
 
$K_1 < K_2$.

In [13]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

0.03768977635909404

In [14]:
european_option = EuropeanOptions(S_t=100,K=75,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

1.0094560974287488

In [15]:
1.0094560974287488 - 0.03768977635909404

0.9717663210696548

$$0 \leq 0.9717663210696548 \leq 25$$
$$0 \leq P(100,0;75,1) - P(100,0;50,1) \leq 25$$

Then, again, the Black and Scholes model respects these put properties

## Call Option 

### Fifth Property 

Consider two European call options $C(S_t, t; K_1, T)$ and $C(S_t, t; K_2, T)$, with strike prices $K_1$ and $K_2$ such that $K_1 < K_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
C(S_t, t; K_1, T) \geq C(S_t, t; K_2, T)
$$
for $K_1 < K_2$.

In [16]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

59.10115212246

In [17]:
european_option = EuropeanOptions(S_t=100,K=75,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

39.604649616580126

In [18]:
59.10115212246 - 39.604649616580126

19.496502505879874

$$39.604649616580126 \geq 19.496502505879874$$
$$C(100,0;50,1) \geq C(100,0;75,1)$$

The call price is then a decreasing function of the strike $K$ . The more the strike, the less the call option price.

## Put Option 

### Fifth Property 

Consider two European put options $P(S_t, t; K_1, T)$ and $P(S_t, t; K_2, T)$, with strike prices $K_1$ and $K_2$ such that $K_1 < K_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
P(S_t, t; K_1, T) \leq P(S_t, t; K_2, T)
$$
for $K_1 < K_2$.

In [19]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

0.03768977635909404

In [20]:
european_option = EuropeanOptions(S_t=100,K=75,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

1.0094560974287488

$$0.03768977635909404 \leq 1.0094560974287488$$
$$P(100,0;50,1) \leq P(100,0;75,1)$$

The put price is then an increasing function of the strike $K$ . The more the strike, the more the put option price .

## Call Option 

### Sixth Property 

Consider three European call options $C(S_t, t; K_1, T)$, $C(S_t, t; K_2, T)$ and $C(S_t, t; K_3, T)$, with strike prices $K_1$, $K_2$ and $K_3$ such that $K_1 < K_2$ < $K_3$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
C(S_t, t; K_1, T) - C(S_t, t; K_2, T) \geq C(S_t, t; K_2, T) - C(S_t, t; K_3, T)
$$
for $K_1 < K_2$ < $K_3$.

### Consequences : $$2 \times C(S_t, t; K_2, T) \leq C(S_t, t; K_1, T) + C(S_t, t; K_3, T)$$

In [21]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

59.10115212246

In [22]:
european_option = EuropeanOptions(S_t=100,K=75,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

39.604649616580126

In [23]:
european_option = EuropeanOptions(S_t=100,K=95,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

26.461028027778546

In [24]:
59.10115212246 - 39.604649616580126 

19.496502505879874

In [25]:
39.604649616580126 - 26.461028027778546

13.14362158880158

In [26]:
59.10115212246 - 39.604649616580126 > 39.604649616580126 - 26.461028027778546

True

In [27]:
2 * 39.604649616580126 

79.20929923316025

In [28]:
59.10115212246 + 26.461028027778546

85.56218015023855

In [29]:
2 * 39.604649616580126 < 85.56218015023855

True

$$59.10115212246 - 39.604649616580126 \geq 39.604649616580126 - 26.461028027778546$$
$$C(100, 0; 50, 1) - C(100, 0; 75, 1) \geq C(100, 0; 75, 1) - C(100, 0; 95, 1)$$

$$2 * 39.604649616580126 \leq 59.10115212246 + 26.461028027778546$$
$$2 \times C(100, 0; 75, 1) \leq C(100, 0; 50, 1) + C(100, 0; 95, 1)$$

The second inequality is very important : It tells us that : 
- 2 calls options prices of strike $K_2$ are always less than 
- One call option price of strike $K_1$ + one call option of strike $K_3$, 

for $K_1 \leq K_2 \leq K_3$ . 

In other words, a call option price is a convex function of the strike price . This inequality is very crucial since it is the core inequality of an option trading strategy called **Call Butterfly Spread**

## Put Option 

### Sixth Property 

Consider three European puts options $P(S_t, t; K_1, T)$, $P(S_t, t; K_2, T)$ and $P(S_t, t; K_3, T)$, with strike prices $K_1$, $K_2$ and $K_3$ such that $K_1 < K_2$ < $K_3$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
P(S_t, t; K_2, T) - P(S_t, t; K_1, T) \leq P(S_t, t; K_3, T) - P(S_t, t; K_2, T)
$$
for $K_1 < K_2$ < $K_3$.

### Consequences : $$2 \times P(S_t, t; K_2, T) \leq P(S_t, t; K_1, T) + P(S_t, t; K_3, T)$$

In [30]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

0.03768977635909404

In [31]:
european_option = EuropeanOptions(S_t=100,K=75,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

1.0094560974287488

In [32]:
european_option = EuropeanOptions(S_t=100,K=95,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

4.240449570186822

In [33]:
1.0094560974287488 - 0.03768977635909404

0.9717663210696548

In [34]:
4.240449570186822 - 1.0094560974287488

3.2309934727580734

In [35]:
2 * 1.0094560974287488

2.0189121948574975

In [36]:
0.03768977635909404 + 4.240449570186822

4.278139346545916

$$1.0094560974287488 - 0.03768977635909404 \leq 4.240449570186822 - 1.0094560974287488$$
$$P(100, 0; 75, 1) - P(100, 0; 50, 1) \leq P(100, 0; 95, 1) - P(100, 0; 75, 1)$$

$$2 * 1.0094560974287488 \leq 0.03768977635909404 + 4.240449570186822$$
$$2 \times P(100, 0; 75, 1) \leq P(100, 0; 50, 1) + P(100, 0; 95, 1)$$

The second inequality is very important : It tells us that : 
- 2 puts options prices of strike $K_2$ are always less than 
- One put option price of strike $K_1$ + one put option of strike $K_3$, 

for $K_1 \leq K_2 \leq K_3$ . 

In other words, a put option price is a convex function of the strike price . This inequality is very crucial since it is the core inequality of an option trading strategy called **Put Butterfly Spread**

## Call Option 

### Seventh Property : American vs European Options


### Notation

Let the following notations be defined:

- $C_{\text{am}}(S_t, t; K, T)$ = Price of the American call option at time $t$,,
- $C_{\text{bs}}(S_t, t; K, T)$ = Price of the European call option at time $t$,
- $S_t$ = Spot price of the underlying dividend-paying stock at time $t$,
- $K$ = Strike price of the options,
- $T$ = Expiry time of the options,
- $r$ = Risk-free interest rate,
- $D$ = Continuous dividend yield.
- $t$ = Pricing time


An **American option** can be exercised at any time before or on its expiration date, whereas a **European option** can only be exercised at expiration. As a result, the value of an American option is generally greater than that of its equivalent European option because the holder of the American option has the additional flexibility to exercise early. Hence : 

   $$
   C_{\text{am}}(S_T, t; K, T) \geq C_{\text{bs}}(S_T, t; K, T)
   $$
  


### American Option and the Black-Scholes Equation

The **Black-Scholes equation** applies only to European options, as these can only be exercised at maturity. Since an American option can be exercised at any point in time before expiration, it does not satisfy the same problem specification as the European option. Therefore, an American option **cannot** satisfy the Black-Scholes equation. However, both American and European options have the **same value at the expiry time**, as they will both settle in the same way at expiration.


### Properties of Europeans Call Options wrt American Call Options

For both American and European call options, the following properties hold:

1. **At Expiry**: The prices of the American and European call options are equal at expiration:

   $$
   C_{\text{am}}(S_T, T; K, T) = C_{\text{bs}}(S_T, T; K, T)
   $$

   This is because both options will have the same payoff at maturity, regardless of whether they are American or European options.

2. **American Call Option**: The value of an American call option is greater than or equal to that of a European call option. This is due to the early exercise feature of American options, which allows the holder to exercise the option before expiry if it becomes advantageous to do so (for example, if the stock pays a dividend).

3. **No Early Exercise in Non-Dividend Paying Stocks**: If the underlying stock does not pay dividends, there is no advantage in exercising the American call option early. In this case, the American call option price will be equivalent to the European call option price.

4. **Dividend-Paying Stocks**: For dividend-paying stocks, early exercise may be beneficial for American call options if the dividend payment exceeds the time value of the option. This makes the American option more valuable compared to the European option, as the holder can choose to exercise the option to capture dividends.

## Put Option 

### Seventh Property : American vs European Options


### Notation

Let the following notations be defined:

- $P_{\text{am}}(S_t, t; K, T)$ = Price of the American put option at time $t$,,
- $P_{\text{bs}}(S_t, t; K, T)$ = Price of the European put option at time $t$,
- $S_t$ = Spot price of the underlying dividend-paying stock at time $t$,
- $K$ = Strike price of the options,
- $T$ = Expiry time of the options,
- $r$ = Risk-free interest rate,
- $D$ = Continuous dividend yield.
- $t$ = Pricing time


An **American option** can be exercised at any time before or on its expiration date, whereas a **European option** can only be exercised at expiration. As a result, the value of an American option is generally greater than that of its equivalent European option because the holder of the American option has the additional flexibility to exercise early. Hence : 

   $$
   P_{\text{am}}(S_T, t; K, T) \geq P_{\text{bs}}(S_T, t; K, T)
   $$
  


### American Option and the Black-Scholes Equation

The **Black-Scholes equation** applies only to European options, as these can only be exercised at maturity. Since an American option can be exercised at any point in time before expiration, it does not satisfy the same problem specification as the European option. Therefore, an American option **cannot** satisfy the Black-Scholes equation. However, both American and European options have the **same value at the expiry time**, as they will both settle in the same way at expiration.


### Properties of Europeans Put Options wrt American Put Options

For both American and European put options, the following properties hold:

1. **At Expiry**: The prices of the American and European put options are equal at expiration:

   $$
   P_{\text{am}}(S_T, T; K, T) = P_{\text{bs}}(S_T, T; K, T)
   $$

   This is because both options will have the same payoff at maturity, regardless of whether they are American or European options.

2. **American Put Option**: The value of an American put option is greater than or equal to that of a European put option. This is due to the early exercise feature of American options, which allows the holder to exercise the option before expiry if it becomes advantageous to do so (for example, if the stock pays a dividend for a call or market pays an interest rate for the put).

3. **No Early Exercise in a no Interest Rate Environment**: If the market does not pay any interest rate, there is no advantage in exercising the American put option early. In this case, the American put option price will be equivalent to the European put option price.

4. **Market Paying Interest Rate**: For market paying interest rate, early exercise may be beneficial for American put options . This makes the American put option more valuable compared to the European put option, as the holder can choose to exercise the option to earn interest rate on strike. 

## European Call - Put Parity

### Problem Statement

At time $t$, we consider a non-dividend-paying stock with spot price $S_t$ and a risk-free interest rate $r$. 

By taking along the European call option price at time $t$, $C(S_t, t; K, T)$, and a short European put option price at time $t$, $P(S_t, t; K, T)$, on the same underlying stock $S_t$, strike price $K$, and expiry time $T$ ($t < T$), we have the following relationship:

$$
C(S_t, t; K, T) - P(S_t, t; K, T) = S_t - K e^{-r(T - t)}
$$

### Explanation

- $C(S_t, t; K, T)$ represents the price of a European call option at time $t$ with strike price $K$ and expiry time \( T \).
- $P(S_t, t; K, T)$ represents the price of a European put option at time $t$ with strike price $K$ and expiry time $T$.
- $S_t$ is the spot price of the stock at time $t$.
- $r$ is the risk-free interest rate.
- $T$ is the expiration time, and $t$ is the current time, where $t < T$.

In [37]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Call_Black_Scholes()

59.10115212246

In [38]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0)
european_option.Put_Black_Scholes()

0.03768977635909404

In [39]:
import numpy as np

59.10115212246 - 0.03768977635909404

59.063462346100906

In [40]:
100 - 50 * np.exp(-0.20 * 1)

59.063462346100906

In [41]:
59.10115212246 - 0.03768977635909404 == 100 - 50 * np.exp(-0.20 * 1)

True

### Verification of Call-Put Parity for European Options

The above relationship is valid for **European options** where both the call and the put options have the same strike price \( K \) and expiry time \( T \), and there are no dividends paid by the underlying asset. It is used to ensure that arbitrage opportunities do not exist in the market.

### Variations for Other Types of Options

While the call-put parity holds for standard European options, this relationship can take different forms depending on the type of option:

1. **Barrier Options**: These options have certain conditions attached to their payoff depending on whether the underlying asset's price hits a certain barrier level during the life of the option. The call-put parity for barrier options would be modified to account for the specific features of the barrier, such as knock-in or knock-out options.

2. **Asian Options**: Asian options depend on the average price of the underlying asset over a specified period rather than the spot price at maturity. The call-put parity for Asian options will incorporate this averaging effect, and the pricing formula will differ from that of European options.

3. **Exotic Options**: Exotic options, which include a wide range of non-standard features (such as options with more complex payoff structures, conditional payoffs, or embedded options), have different pricing models. As a result, the call-put parity for exotic options will vary depending on the specific characteristics of the exotic contract.

4. **American Options** where we must choose the optimal moment to exercise or not the option . The Call-Put parity may very much ( inequalities ) . 


In [42]:
import warnings
warnings.filterwarnings(action='ignore')

## Asymptotics and Limits of European Options Price 


In this section, we will see how the price of an european option ( Call or Put ) behaves with respect to the Black & Scholes Parameters and Variables . 

### Call Option

Let $C(S_t, t; K, T)$ represents the price of a call option with underlying spot price $S_t$ at time $t$

where:
- $K$ is the strike price of the option,
- $T$ is the expiry time of the option,
- $D$ is the continuous dividend yield,
- $e^{-D \cdot (T - t)}$ is the discount factor due to the dividend yield over the remaining time until expiry, $T - t$.


**Underlying Asset limit $S_t$**


$$
\lim_{{S_t \to \infty}} C(S_t, t; K, T) = S_t \cdot e^{-D \cdot (T - t)}
$$


and 

$$
\lim_{{S_t \to 0}} C(S_t, t; K, T) = 0
$$

This limit indicates that as the spot price $S_t$ increases without bound, the call option price converges to the present value of $S_t$, adjusted for the dividend yield and towards 0 when the spot price becomes smaller and smaller . 

Let's prove that using our Black & Scholes Pricer 

In [43]:
european_option = EuropeanOptions(S_t=1000000,K=50,T=1,sigma=0.35,r=0.2,D=0) #S_t = 10^6
european_option.Call_Black_Scholes()

999959.0634623461

We see that when we take the spot price equals to $1000000$, then price of our option is $999959.0634623461$ which is very close to $1000000$ . In fact, if we take the spot price bigger and bigger, the european call price will be close to the spot price . Our dividend yield is $0$. So, the difference between our pricer result and our spot price is due to how high the price of the spot we input in our pricer is.

In [44]:
european_option = EuropeanOptions(S_t=0,K=50,T=1,sigma=0.35,r=0.2,D=0) #S_t = 0
european_option.Call_Black_Scholes()

0.0

We see also that if the spot goes to $0$, then the european call option goes to $0$. Intuitively, the option will never be in the money .

**Strike limit $K$**

$$
\lim_{{K \to \infty}} C(S_t, t; K, T) = 0
$$


and 

$$
\lim_{{S_t \to 0}} C(S_t, t; K, T) = S_t \cdot e^{-D \cdot (T - t)}
$$


In [45]:
european_option = EuropeanOptions(S_t=100,K=5000000000,T=1,sigma=0.35,r=0.2,D=0) #K= 5*10^9
european_option.Call_Black_Scholes()

0.0

In [46]:
european_option = EuropeanOptions(S_t=100,K=0.000000001,T=1,sigma=0.35,r=0.2,D=0) #K= 10^-8
european_option.Call_Black_Scholes()

99.99999999918127

We can see that effectively, when the strike $K$ goes to $\infty$, then our call price goes towards $0$ and when the strike $K$ goes to $0$, then the call price goes to the spot . 

**Volatility limit $\sigma$**

$$
\lim_{{\sigma \to \infty}} C(S_t, t; K, T) = S_t \cdot e^{-D \cdot (T - t)}
$$


and 

$$
\lim_{\sigma \to 0} \text{C}(S_t, t, K, T) = 
\begin{cases}
    S_t \cdot e^{-D (T - t)} - K \cdot e^{-r (T - t)} & \text{if } S_t > K \cdot e^{-(r - D)(T - t)} \\
    0 & \text{if } S_t \leq K \cdot e^{-(r - D)(T - t)}
\end{cases}
$$

In [47]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=20,r=0.2,D=0) #sigma = 2000%
european_option.Call_Black_Scholes()

100.0

In [48]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0,r=0.2,D=0) #sigma = 0
european_option.Call_Black_Scholes()

59.063462346100906

Since the limit when $\sigma$ goes to $0$ is not nul, then necessarily, according to our volatility limit function, then the call option price takes its value in the first case of the function if $S_t > K \cdot e^{-(r - D)(T - t)}$

$S_t \cdot e^{-D (T - t)} - K \cdot e^{-r (T - t)}$ is then equals to 

In [49]:
100 - 50 * np.exp(-0.2 * 1) # according to the volatility function for sigma = 0 

59.063462346100906

which confirms our result

**Dividend Yield limit $D$**

$$
\lim_{{D \to \infty}} C(S_t, t; K, T) = 0
$$


$$
\lim_{{D \to 0}} C(S_t, t; K, T) = S_t \Phi(d_+) - K e^{-r(T - t)} \Phi(d_-)
$$

In [50]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.2,D=20) #D = 2000%
european_option.Call_Black_Scholes()

0.0

In [51]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.2,D=0) #D = 0
european_option.Call_Black_Scholes()

59.10115212246

**Maturity Limit $T-t$**
 
$$
\lim_{{t \to T}} C(S_t, t; K, T) = max(S_T - K, 0)
$$ 
 

$$
\lim_{T-t \to \infty, \sigma > 0} \text{C}(S_t, t, K, T) = 
\begin{cases}
    S_t & \text{if } D = 0 \\
    0 & \text{if } D > 0
\end{cases}
$$

In [52]:
european_option = EuropeanOptions(S_t=100,K=50,T=1000,sigma=0.35,r=0.2,D=0) #T = 1000 years , D = 0
european_option.Call_Black_Scholes()

100.0

In [53]:
european_option = EuropeanOptions(S_t=100,K=50,T=1000,sigma=0.35,r=0.2,D=0.7) #T = 1000 years , D > 0 
european_option.Call_Black_Scholes()

0.0

### Put Option

Let $P(S_t, t; K, T)$ represents the price of a call option with underlying spot price $S_t$ at time $t$

where:
- $K$ is the strike price of the option,
- $T$ is the expiry time of the option,
- $D$ is the continuous dividend yield,
- $e^{-D \cdot (T - t)}$ is the discount factor due to the dividend yield over the remaining time until expiry, $T - t$.


**Underlying Asset limit $S_t$**


$$
\lim_{{S_t \to \infty}} P(S_t, t; K, T) = 0
$$


and 

$$
\lim_{{S_t \to 0}} P(S_t, t; K, T) = K \cdot e^{-r \cdot (T - t)}
$$

This limit indicates that as the spot price $S_t$ increases without bound, the put option price converges to $0$, and towards the strike price adjusted to the interest rate when the spot price becomes smaller and smaller . 

Let's prove that using our Black & Scholes Pricer .

In [54]:
european_option = EuropeanOptions(S_t=100000000,K=50,T=1,sigma=0.35,r=0.20,D=0) #S_t = 10^8
european_option.Put_Black_Scholes()

0.0

In [55]:
european_option = EuropeanOptions(S_t=0,K=50,T=1,sigma=0.35,r=0.20,D=0) #S_t = 0 
european_option.Put_Black_Scholes()

40.936537653899094

In [56]:
50 * np.exp(-0.2 * 1)

40.936537653899094

$40.93 \approx 50$ which is our strike price . So our pricer is good and valid

**Strike limit $K$**

$$
\lim_{{K \to \infty}} P(S_t, t; K, T) = K \cdot e^{-r \cdot (T - t)}
$$


and 

$$
\lim_{{K \to 0}} P(S_t, t; K, T) = 0
$$

In [57]:
european_option = EuropeanOptions(S_t=100,K=5000000000,T=1,sigma=0.35,r=0.20,D=0) #K = 5*10^9
european_option.Put_Black_Scholes()

4093653665.3899093

In [58]:
5000000000 * np.exp(-0.20*1)

4093653765.3899093

In [59]:
european_option = EuropeanOptions(S_t=100,K=0.00001,T=1,sigma=0.35,r=0.20,D=0) #K = 10^-4
european_option.Put_Black_Scholes()

0.0

$4093653665.3899093 \approx 5000000000$. Our pricer is ok and correct

**Volatility limit $\sigma$**


$$
\lim_{{\sigma \to \infty}} P(S_t, t; K, T) = K \cdot e^{-r \cdot (T - t)}
$$


and 

$$
\lim_{\sigma \to 0} \text{P}(S_t, t, K, T) = 
\begin{cases}
     K \cdot e^{-r (T - t)} - S_t \cdot e^{-D (T - t)} & \text{if } S_t < K \cdot e^{-(r - D)(T - t)} \\
    0 & \text{if } S_t \geq K \cdot e^{-(r - D)(T - t)}
\end{cases}
$$

In [60]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=20,r=0.20,D=0) #sigma = 2000%
european_option.Put_Black_Scholes()

40.936537653899094

In [61]:
50 * np.exp(-0.2 * 1)

40.936537653899094

In [62]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0,r=0.20,D=0) #sigma = 0%
european_option.Put_Black_Scholes()

0.0

Since the limit when $\sigma$ goes to $0$ is nul, then necessarily, according to our volatility limit function, then the put option price takes its value in the first case of the function if $S_t < K \cdot e^{-(r - D)(T - t)}$. So, the value of the put option is $0$.

**Dividend Yield limit $D$**


$$
\lim_{{D \to \infty}} P(S_t, t; K, T) = K \cdot e^{-r \cdot (T - t)}
$$


$$
\lim_{{D \to 0}} P(S_t, t; K, T) = K e^{-r(T - t)} \Phi(-d_-) - S_t \Phi(-d_+)
$$

In [63]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=20) #D = 2000%
european_option.Put_Black_Scholes()

40.936537447783735

In [64]:
50 * np.exp(-0.2 * 1)

40.936537653899094

In [65]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0) #D = 0%
european_option.Put_Black_Scholes()

0.03768977635909404

**Maturity Limit $T-t$**
 
$$
\lim_{{t \to T}} P(S_t, t; K, T) = max(K - S_T, 0)
$$ 
 

$$
\lim_{T-t \to \infty, \sigma > 0} \text{P}(S_t, t, K, T) = 
\begin{cases}
    K & \text{if } r = 0 \\
    0 & \text{if } r > 0
\end{cases}
$$

In [66]:
european_option = EuropeanOptions(S_t=100,K=50,T=1000,sigma=0.35,r=0,D=0) #T = 1000 years , r = 0 
european_option.Put_Black_Scholes()

49.99999779060548

$49.99999779060548 \approx 50$ if $r = 0$

In [67]:
european_option = EuropeanOptions(S_t=100,K=50,T=1000,sigma=0.35,r=0.7,D=0) #T = 1000 years , r > 0 
european_option.Put_Black_Scholes()

0.0

## Call Option 

### Eighth Property 

Consider two European call options $C(S_t, t; K, T,r_1)$ and $C(S_t, t; K, T, r_2)$, with expiry times $r_1$ and $r_2$ such that $r_1 < r_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
C(S_t, t; K, T, r_1) \leq C(S_t, t; K, T, r_2)
$$
for $r_1 < r_2$.

In [68]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0) #r=0.2
european_option.Call_Black_Scholes()

59.10115212246

In [69]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.40,D=0) #r=0.4
european_option.Call_Black_Scholes()

66.4889079323494

$$59.10115212246 \leq 66.4889079323494$$
$$C(100, 0; 50, 1, 0.20) \leq C(100, 0; 50, 1, 0.40)$$

## Put Option 

### Eighth Property 

Consider two European call options $P(S_t, t; K, T,r_1)$ and $P(S_t, t; K, T, r_2)$, with expiry times $r_1$ and $r_2$ such that $r_1 < r_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
P(S_t, t; K, T, r_1) \geq P(S_t, t; K, T, r_2)
$$
for $r_1 < r_2$.

In [70]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.2,D=0) #r=0.2 
european_option.Put_Black_Scholes()

0.03768977635909404

In [71]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.40,D=0) #r=0.4
european_option.Put_Black_Scholes()

0.0049102341313603665

$$0.03768977635909404 \geq 0.0049102341313603665$$
$$P(100, 0; 50, 1, 0.20) \geq P(100, 0; 50, 1, 0.40)$$

## Call Option 

### Nineth Property 

Consider two European call options $C(S_t, t; K, T,\sigma_1)$ and $C(S_t, t; K, T, \sigma_2)$, with expiry times $\sigma_1$ and $\sigma_2$ such that $\sigma_1 < \sigma_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
C(S_t, t; K, T, \sigma_1) \leq C(S_t, t; K, T, \sigma_2)
$$
for $\sigma_1 < \sigma_2$.


In fact, this property is just another form of the limit of a call option price when $\sigma$ goes to $\infty$ since the call option price increases with volatility.

In [72]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0) #sigma=0.35
european_option.Call_Black_Scholes()

59.10115212246

In [73]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.55,r=0.20,D=0) #sigma=0.55
european_option.Call_Black_Scholes()

59.815019258245925

$$59.10115212246 < 59.815019258245925$$
$$C(100, 0; 50, 1, 0.35) \leq C(100, 0; 50, 1, 0.55)$$

## Put Option 

### Nineth Property 

Consider two European put options $P(S_t, t; K, T,\sigma_1)$ and $P(S_t, t; K, T, \sigma_2)$, with expiry times $\sigma_1$ and $\sigma_2$ such that $\sigma_1 < \sigma_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
P(S_t, t; K, T, \sigma_1) \leq P(S_t, t; K, T, \sigma_2)
$$
for $\sigma_1 < \sigma_2$.


In fact, this property is just another form of the limit of a put option price when $\sigma$ goes to $\infty$ since the call option price increases with volatility.

In [74]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.2,D=0) #sigma=0.35 
european_option.Put_Black_Scholes()

0.03768977635909404

In [75]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.55,r=0.2,D=0) #sigma=0.55 
european_option.Put_Black_Scholes()

0.7515569121450127

$$0.03768977635909404 < 0.7515569121450127$$
$$P(100, 0; 50, 1, 0.35) \leq P(100, 0; 50, 1, 0.55)$$

## Call Option 

### Tenth Property 

Consider two European call options $C(S_t, t; K, T,D_1)$ and $C(S_t, t; K, T, D_2)$, with expiry times $D_1$ and $D_2$ such that $D_1 < D_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
C(S_t, t; K, T, D_1) \geq C(S_t, t; K, T, D_2)
$$
for $D_1 < D_2$.


In fact, this property is just another form of the limit of a call option price when $D$ goes to $\infty$ since the call option price decreases with dividend.

In [76]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0.2) #D=0.2
european_option.Call_Black_Scholes()

41.115655851784574

In [77]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.20,D=0.4) #D=0.4
european_option.Call_Black_Scholes()

26.747433433973526

$$41.115655851784574 > 26.747433433973526$$
$$C(100, 0; 50, 1, 0.2) \geq C(100, 0; 50, 1, 0.4)$$

## Put Option 

### Tenth Property 

Consider two European call options $P(S_t, t; K, T,D_1)$ and $P(S_t, t; K, T, D_2)$, with expiry times $D_1$ and $D_2$ such that $D_1 < D_2$. Here:
- $S_t$ is the spot price of the underlying asset at time $t$,
- $K$ is the common strike price.

Then : 
$$
P(S_t, t; K, T, D_1) \leq P(S_t, t; K, T, D_2)
$$
for $D_1 < D_2$.


In fact, this property is just another form of the limit of a put option price when $D$ goes to $\infty$ since the put option price increases with dividend.

In [78]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.2,D=0.2) #D=0.2 
european_option.Put_Black_Scholes()

0.17911819788547945

In [79]:
european_option = EuropeanOptions(S_t=100,K=50,T=1,sigma=0.35,r=0.2,D=0.4) #D=0.4
european_option.Put_Black_Scholes()

0.6519664843086925

$$0.17911819788547945 < 0.6519664843086925$$
$$P(100, 0; 50, 1, 0.2) \leq P(100, 0; 50, 1, 0.4)$$